ui/client/components/graph/SelectionContextProvider.tsx

```diff
@@ -0,0 +1,226 @@
+  copy: UserAction,
+  paste: UserAction,
+  cut: UserAction,
+  delete: UserAction,
+  undo: UserAction,
+  redo: UserAction,
+  selectAll: UserAction,
+}
+
+function useClipboardParse() {
+  const processCategory = useSelector(getProcessCategory)
+  const processDefinitionData = useSelector(getProcessDefinitionData)
+  return useCallback(
+    text => {
+      const selection = JsonUtils.tryParseOrNull(text)
+      const isValid = selection?.edges &&
+        selection?.nodes?.every(node => NodeUtils.isNode(node) &&
+          NodeUtils.isPlainNode(node) &&
+          NodeUtils.isAvailable(node, processDefinitionData, processCategory))
+      return isValid ? selection : null
+    },
+    [processCategory, processDefinitionData],
+  )
+}
+
+function useClipboardPermission(): boolean | string {
+  const clipboardPermission = useRef<PermissionStatus>()
+  const [state, setState] = useState<"denied" | "granted" | "prompt">()
+  const [text, setText] = useState("")
+  const [content, setContent] = useState("")
+
+  const parse = useClipboardParse()
+
+  if (state === "granted") {
+    navigator.clipboard.readText().then(text => {
+      setText(text)
+    }).catch(() => {return})
+  }
+
+  useEffect(() => {
+    setContent(parse(text))
+  }, [text])
+
+  useEffect(() => {
+    navigator.permissions.query({name: "clipboard-read"}).then(permission => {
+      clipboardPermission.current = permission
+      setState(permission.state)
+      permission.onchange = () => {
+        setState(permission.state)
+      }
+    })
+    return () => {
+      if (clipboardPermission.current) {
+        clipboardPermission.current.onchange = undefined
+      }
+    }
+  }, [])
+
+  return state === "prompt" || content
+}
+
+const SelectionContext = createContext<UserActions>(null)
+
+export const useSelectionActions = (): UserActions => {
+  const selectionActions = useContext(SelectionContext)
+  if (!selectionActions) {
+    throw new Error("used useSelectionActions outside provider")
+  }
+  return selectionActions
+}
+
+export default function SelectionContextProvider(props: PropsWithChildren<{ pastePosition: () => { x: number, y: number } }>): JSX.Element {
+  const dispatch = useDispatch()
+  const {t} = useTranslation()
+
+  const selectionState = useSelector(getSelectionState)
+  const capabilities = useSelector(getCapabilities)
+  const selection = useSelector(getSelection)
+  const canModifySelected = useSelector(canModifySelectedNodes)
+
+  const [hasSelection, setHasSelection] = useState(hasTextSelection)
+
+  const copy = useCallback(
+    async (silent = false) => {
+      if (silent && hasTextSelection()) {
+        return
+      }
+
+      if (canModifySelected) {
+        await ClipboardUtils.writeText(JSON.stringify(selection))
+        const {nodes} = selection
+        if (!silent) {
+          dispatch(success(t("userActions.copy", {
+            defaultValue: "Copied node",
+            defaultValue_plural: "Copied {{count}} nodes",
```

Comment:
Two styles of formatting..? Maybe defaultValuePlural?

Reply:
what? please read about i18next.